DeUrgenta/Features/Authentication/SignIn/ViewModel/SignInViewModel.swift

```diff
@@ -0,0 +1,52 @@
+import Foundation
+
+class SignInViewModel: ObservableObject {
+    @Published var email: String = ""
+    @Published var password: String = ""
+    @Published var errorMessaageEmail: String = ""
+    @Published var errorMessaagePassword: String = ""
+    @Published var showEmailError: Bool = false
+    @Published var showPasswordError: Bool = false
+
+    private let validator = Validator.shared
+
+    func isValidEmail() -> Bool {
+        showEmailError = false
+
+        validator.isValidData(data: email, type: .email) { error in
+            guard let error = error else {
+                return
+            }
+            self.showEmailError = true
+
+            switch error {
+            case .empty:
+                self.errorMessaageEmail = AppStrings.Authentication.Error.emptyEmail.localized()
+            case .invalid:
+                self.errorMessaageEmail = AppStrings.Authentication.Error.invalidEmail.localized()
+            }
+        }
+
+        return !showEmailError
+    }
+
+    func isValidPassword() -> Bool {
+        showPasswordError = false
+
```

Comment:
This code is a little bit confusing for me. It is a synchronous function? Because `validator.isValidData ` is asynchronous and this may lead to some data racing issues. The return of the function is not waiting for the isValidData response.